man/rmdhunks/vcr-design.Rmd

```diff
@@ -0,0 +1,152 @@
+hijacking internals of the HTTP clients. 
+
+However, monkey patching is not allowed in R. Thus, in R we have to
+somehow have "hooks" into HTTP clients in R. Fortunately, Scott is the
+maintainer of one of the HTTP clients, `crul`, so was able to quickly
+create a hook. Very fortunately, there was already a hook mechanism
+in the `httr` package. 
+
+The actual hooks are not in `vcr`, but in `webmockr`. `vcr` depends on
+`webmockr` for hooking into HTTP clients `httr` and `crul`. 
+
+## Internal classes
+
+An overview of some of the more important aspects of vcr.
+
+### Configuration
+
+An internal object (`vcr_c`) is created when `vcr` is loaded with
+the default vcr configuration options inside of an R6 class `VCRConfig` -
+see <https://github.com/ropensci/vcr/blob/master/R/onLoad.R>. This
+class is keeps track of default and user specified configuration options.
+You can access `vcr_c` using triple namespace `:::`, though it is not
+intended for general use. Whenever you make calls to `vcr_configure()`
+or other configuration functions, `vcr_c` is affected.
+
+### Cassette class
+
+`Cassette` is an R6 class that handles internals/state for each cassette.
+Each time you run `use_cassette()` this class is used. The class has quite
+a few methods in it, so there's a lot going on in the class. Ideally
+the class would be separated into subclasses to handle similar sets
+of logic, but there's not an easy way to do that with R6. 
+
+Of note in `Cassette` is that when called, within the `initialize()` 
+call `webmockr` is used to create webmockr stubs. 
+
+### How HTTP requests are handled
+
+Within `webmockr`, there's calls to the vcr class `RequestHandler`, which
+has child classes `RequestHandlerCrul` and `RequestHandlerHttr` for 
+`crul` and `httr`, respectively. These classes determine what to do with
+each HTTP request. The options for each HTTP request include:
+
+- **Ignored** You can ignore HTTP requests under certain rules using the 
+configuration options `ignore_hosts` and `ignore_localhost`
+- **Stubbed by vcr** This is an HTTP request for which a match is found
+in the cassette defined in the `use_cassette()`/`insert_cassette()` call.
+In this case the matching request/response from the cassette is returned
+with no real HTTP request allowed.
+- **Recordable** This is an HTTP request for which no match is found
+in the cassette defined in the `use_cassette()`/`insert_cassette()` call.
+In this case a real HTTP request is allowed, and the request/response is
+recorded to the cassette.
+- **Unhandled** This is a group of cases, all of which cause an error
+to be thrown with a message trying to help the user figure out how to
+fix the problem.
+
+If you use vcr logging you'll see these categories in your logs.
+
+### Serializers
+
+Serializers handle in what format cassettes are written to files on disk.
+The current options are YAML (default) and JSON. 
+
+An R6 class `Serializer` is the parent class for all serializer types;
```

Comment:
Maybe add context on why you implemented YAML first and the problems of YAML?

Reply:
good idea